### PR TITLE
Dk/uint backend

### PR DESCRIPTION
**Pull request by rw0x0**
> [!NOTE]
> **High Risk**
> Breaking MPC binary-share serialization and widespread changes to comparison, bitwise, and witness-generation paths across Circom and Noir stacks; coordinated upgrade required and correctness hinges on per-field width contracts.
> 
> **Overview**
> **Replaces REP3 binary-domain `BigUint` with stack-allocated fixed-width integers** via new `mpc_core::uint` (`UintBackend`, `FieldUint`, `RUint`) and renames **`Rep3BigUintShare` → `Rep3UintShare`**, wiring a2b/b2a, comparisons, bitwise ops, and RNG paths through `to_uint` / `from_uint_reduced` instead of `num-bigint`.
> 
> **Protocol / semantics changes in mpc-core:** binary shares serialize as **fixed-size** little-endian bytes (not variable-length `BigUint`); **`shift_l_public`** panics when shift ≥ modulus bit size; **`binary::and`** masking/randomness uses full uint width (odd-modulus `is_zero` padding fix); edge case for **`unsigned_ge_const_rhs` with zero** is fixed.
> 
> **Downstream (co-circom / circom-mpc-vm, co-acvm, co-brillig, co-noir-*)** adds **`F: FieldUint`** (and paired base-field bounds where needed) on witness/VM/prover code, swaps **`Rep3UintShare`** and uint bit helpers in plain/REP3 drivers, and trims several **`num-bigint`** / dev-only crate dependencies. **`docs/uint-backend-migration.md`** documents the migration and status.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9999044ac52c7ab5db66e8fe1d55e553f90fba56. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: mpc-core/src/protocols/rep3/port_tests.rs [hunk 1]
```diff
@@ -0,0 +1,684 @@
...
+    }
+}
+
+/// Test 9b: Required edge case from the task brief: `unsigned_ge_const_rhs(x, 0)`
+/// must be 1 for every `x` (every field element is >= 0). This is split out
+/// from `case_unsigned_ge_semantics` so that, if it fails against the
+/// current implementation, only this case needs `#[ignore]`, without
+/// weakening the rest of the suite's assertions.
+///
+/// CONFIRMED PRE-EXISTING BUG (both bn254 and bls12-381): `y=0` makes
```
Comment (rw0x0) [inline mpc-core/src/protocols/rep3/port_tests.rs]: Is this ever fixed?

Comment (rw0x0) [reply mpc-core/src/protocols/rep3/port_tests.rs]: indeed it seems to be

File: mpc-core/src/protocols/rep3/detail.rs [hunk 2]
```diff
@@ -73,146 +245,177 @@ fn kogge_stone_inner_many<F: PrimeField, N: Network>(
...
-        *y &= &mask;
+        y.and_mask_assign(&mask);
     }
-    // Spread the ov share to the whole biguint
+    // Spread the ov share to the whole uint
     let mut ov = Vec::with_capacity(y_msb.len());
     for y_msb in y_msb {
-        let ov_a = if y_msb.a.iter_u64_digits().next().unwrap_or_default() & 1 == 1 {
-            mask.clone()
+        let ov_a = if y_msb.a.to_u64_truncating() & 1 == 1 {
```
Comment (rw0x0) [inline mpc-core/src/protocols/rep3/detail.rs]: this could be .bit(0)

File: mpc-core/src/protocols/rep3/port_tests.rs [hunk 3]
```diff
@@ -11,9 +11,15 @@
 //!   construct or inspect the raw share representation directly.
 //!
 //! Run with: `cargo test -p mpc-core --all-features --lib port_tests`
+//!
+//! The share type is now `Copy`; the pre-port test bodies (which must stay
```
Comment (rw0x0) [inline mpc-core/src/protocols/rep3/port_tests.rs]: we should fix this at some point

File: mpc-core/src/protocols/rep3/port_tests.rs [hunk 4]
```diff
@@ -24,24 +30,47 @@ const SEED: u64 = 0xdead_beef;
...
 /// The ONLY module the uint-port task may touch. Converts between the test
-/// bodies' canonical types (`F`, `BigUint`, `bool`) and the current
-/// `BigUint`-backed share representation (`Rep3BigUintShare<F>`). Every spot
+/// bodies' canonical types (`F`, `BigUint`, `bool`) and the fixed-width
+/// `F::Uint`-backed share representation (`Rep3UintShare<F>`). Every spot
 /// in this file that directly reads/writes a share's `.a`/`.b` fields, or
 /// constructs a "public mask" value in the share's native domain, lives here.
+///
+/// Since the ported production API speaks `F::Uint` at the public-mask and
+/// combine boundaries while the test bodies keep their canonical `BigUint`
```
Comment (rw0x0) [inline mpc-core/src/protocols/rep3/port_tests.rs]: same here

File: mpc-core/src/protocols/rep3_ring/ring/int_ring.rs [hunk 5]
```diff
@@ -313,4 +372,12 @@ impl<const BITS: usize, const LIMBS: usize> IntRing2k for RUint<BITS, LIMBS> {
         }
         Self(ruint::Uint::from_limbs(limbs))
     }
+
+    fn cast_to_uint<U: UintBackend>(&self) -> U {
```
Comment (rw0x0) [inline mpc-core/src/protocols/rep3_ring/ring/int_ring.rs:305]: shouldn't these 2 just be noops?

File: mpc-core/src/protocols/rep3.rs [hunk 6]
```diff
@@ -534,21 +534,19 @@ where
     [a, b, c]
 }
 
-/// Secret shares a field element using replicated secret sharing and the provided random number generator. The field element is split into three binary shares, where each party holds two. The outputs are of type [Rep3BigUintShare].
-pub fn share_biguint<F: PrimeField, R: Rng + CryptoRng>(
+/// Secret shares a field element using replicated secret sharing and the provided random number generator. The field element is split into three binary shares, where each party holds two. The outputs are of type [Rep3UintShare].
+pub fn share_biguint<F: FieldUint, R: Rng + CryptoRng>(
```
Comment (rw0x0) [inline mpc-core/src/protocols/rep3.rs]: shall we call this also share_uint?

Comment (rw0x0) [reply mpc-core/src/protocols/rep3.rs]: is done in a later commit

File: co-noir/co-acvm/src/mpc/rep3.rs [hunk 7]
```diff
@@ -355,7 +354,60 @@ fn get_base_powers<const NUM_SLICES: usize>(base: u64) -> [BigUint; NUM_SLICES]
...
+///
+/// This mirrors `mpc_core::protocols::rep3::conversion::bit_inject`, but only
+/// requires `F: PrimeField`. It exists because `set_point_to_value_if_zero`'s
+/// generic curve parameter `C` carries no `FieldUint` bound on `C::ScalarField`
+/// in the `NoirWitnessExtensionProtocol` trait signature (which must not
+/// change), so the scalar-field arithmetic share for `pointshare::is_zero`'s
+/// `(bool, bool)` result cannot be produced by wrapping into a
+/// `Rep3UintShare<C::ScalarField>` and calling the (now `FieldUint`-bounded)
+/// generic `bit_inject`.
+fn bool_pair_to_arithmetic_share<F: PrimeField, N: Network>(
```
Comment (rw0x0) [inline co-noir/co-acvm/src/mpc/rep3.rs]: The existance of this function here is weird

File: co-noir/co-acvm/src/mpc/rep3.rs [hunk 8]
```diff
@@ -2358,26 +2412,44 @@ impl<'a, F: PrimeField, N: Network> NoirWitnessExtensionProtocol<F> for Rep3Acvm
     }
 
     // TACEO TODO: we could make the equality check return the basefield element here
+    //
+    // NOTE (uint-backend port): `arithmetic::eq`/`eq_public`/`a2b_selector` now
+    // route through the fixed-width binary domain and require `C::BaseField:
+    // FieldUint`. This trait method only bounds `C::BaseField: PrimeField`
+    // (see `NoirWitnessExtensionProtocol::equals_other_acvm_type` in
+    // `mpc.rs`, which must not change), so that bound cannot be added here
```
Comment (rw0x0) [inline co-noir/co-acvm/src/mpc/rep3.rs:2366]: also these workarounds feel strange

File: co-noir/co-acvm/src/mpc/rep3.rs [hunk 9]
```diff
@@ -2386,11 +2458,17 @@ impl<'a, F: PrimeField, N: Network> NoirWitnessExtensionProtocol<F> for Rep3Acvm
         let res_bin = conversion::a2b_selector(res, self.net0, &mut self.state0)?;
 
         let res_fr = conversion::b2a(
-            &Rep3BigUintShare::new(res_bin.a, res_bin.b),
+            &Rep3UintShare::<ark_bn254::Fr>::new(res_bin.a, res_bin.b),
             self.net0,
             &mut self.state0,
         )?;
-        Ok((res_fr.into(), res.into()))
+        // Safety: we checked that the types match
```
Comment (rw0x0) [inline co-noir/co-acvm/src/mpc/rep3.rs:2411]: same here

File: co-noir/co-acvm/src/mpc/rep3.rs (hunk 7 above)
Comment (rw0x0) [reply co-noir/co-acvm/src/mpc/rep3.rs]: ok is gone in a later commit

File: mpc-core/src/protocols/rep3_ring/yao.rs [hunk 10]
```diff
@@ -980,6 +980,19 @@ where
     )
 }
 
+/// Casts a public divisor (a `BigUint` that may be wider than `T`, e.g. a
+/// field modulus) into a ring element, truncating to `T`'s width like the
+/// former `IntRing2k::cast_from_biguint` did. Routed through [`U512`] as a
+/// wide-enough intermediate since the divisors used here (field moduli) fit
+/// comfortably within 512 bits.
+fn public_divisor_to_ring<T: IntRing2k>(divisor: &BigUint) -> RingElement<T> {
```
Comment (rw0x0) [inline mpc-core/src/protocols/rep3_ring/yao.rs]: I feel like these BigUInts should also be removed

File: tests/src/lib.rs [hunk 11]
```diff
@@ -1,3 +1,12 @@
+/// Converts a fixed-width uint back to a [`num_bigint::BigUint`] for tests
+/// that still compare against an independently computed `BigUint`
+/// expectation.
+pub fn uint_to_biguint<U: mpc_core::uint::UintBackend>(u: &U) -> num_bigint::BigUint {
```
Comment (rw0x0) [inline tests/src/lib.rs]: not sure if we still need it. Imo BigUint should be fully removed

File: tests/tests/mpc/rep3_ring.rs [hunk 12]
```diff
@@ -35,6 +36,22 @@ mod ring_share {
     use rand::Rng;
     use std::sync::mpsc;
 
+    /// Bridges a ring element into its canonical `BigUint` value, for tests
+    /// that still perform independent public-value math (division by a
+    /// field modulus, bit-slicing, etc.) against it. `U512` is the widest
+    /// ring type exercised in this file, so it is always wide enough to
+    /// hold any `T` used here without loss.
+    fn ring_to_biguint<T: IntRing2k>(x: &T) -> BigUint {
```
Comment (rw0x0) [inline tests/tests/mpc/rep3_ring.rs]: same here, should probably be dropped

**Comment by rw0x0** [comment]
cargo machete shows some imports can be removed

File: co-circom/circom-mpc-vm/src/mpc/plain.rs [hunk 13]
```diff
@@ -163,15 +161,16 @@ impl<F: PrimeField> VmCircomWitnessExtension<F> for CircomPlainVmWitnessExtensio
...
+        // result < p since shifting right only shrinks the value
+        Ok(F::from_uint_unchecked(&(val >> shift)))
     }
 
     fn shift_l(&mut self, a: Self::VmType, b: Self::VmType) -> Result<Self::VmType> {
-        let val = to_bigint!(a);
+        // (a << shift) mod p == a * 2^shift mod p, without wide intermediates
         let shift = to_usize!(b);
-        Ok(F::from(val << shift))
+        Ok(a * F::from(2u64).pow([shift as u64]))
```
Comment (rw0x0) [inline co-circom/circom-mpc-vm/src/mpc/plain.rs:173]: This changes behaviour, was probably wrong before but still worth double checking

Comment (dkales) [reply co-circom/circom-mpc-vm/src/mpc/plain.rs:173]: Yeah I thought about this as well, but it actually does not change behavior as this always shifted to larger data types in the biguint and then reduced in the from, so this is equivalent.

File: mpc-core/src/protocols/rep3/detail.rs (hunk 2 above)
Comment (dkales) [reply mpc-core/src/protocols/rep3/detail.rs]: Done — applied `.bit(0)` at all 4 LSB-extraction sites in this file (will be in the next push).

File: mpc-core/src/protocols/rep3/port_tests.rs (hunk 3 above)
Comment (dkales) [reply mpc-core/src/protocols/rep3/port_tests.rs]: Moot now — port_tests.rs was deleted in a later commit.

File: mpc-core/src/protocols/rep3/port_tests.rs (hunk 4 above)
Comment (dkales) [reply mpc-core/src/protocols/rep3/port_tests.rs]: Moot now — port_tests.rs was deleted in a later commit.

File: mpc-core/src/protocols/rep3_ring/ring/int_ring.rs (hunk 5 above)
Comment (dkales) [reply mpc-core/src/protocols/rep3_ring/ring/int_ring.rs:305]: They can't be literal noops: `U: UintBackend` is generic and may have a different width than `Self` (e.g. ring `U512` vs field uint), so the casts must zero-extend/truncate. When `U == Self` the limb copy compiles to a plain move, so the same-type case effectively is a noop.

File: co-noir/co-acvm/src/mpc/rep3.rs (hunk 8 above)
Comment (dkales) [reply co-noir/co-acvm/src/mpc/rep3.rs:2366]: Agreed it's ugly, but I measured the alternative: adding `F: FieldUint + C::BaseField: FieldUint<Uint = F::Uint>` to the trait method propagates through `BigField::equals`/`internal_div` → the `div` wrappers → `BigGroup::add`/`sub`/ladder → essentially the whole big_field/big_group generic surface. Same TypeId pattern as the existing cross-curve helpers here (e.g. `multi_scalar_mul`), so I'd keep it and clean up the whole pattern in one dedicated pass if we ever generalize beyond bn254.

File: co-noir/co-acvm/src/mpc/rep3.rs (hunk 9 above)
Comment (dkales) [reply co-noir/co-acvm/src/mpc/rep3.rs:2411]: Same reasoning as above — generalizing needs the FieldUint width-equality bound on the trait method, which ripples through the whole big_field/big_group API.

File: tests/src/lib.rs (hunk 11 above)
Comment (dkales) [reply tests/src/lib.rs]: The `uint_to_biguint` helper is deleted in a later commit. The only BigUint left here is hex-string input parsing in `split_input_plain` — same deliberately-kept parsing boundary as `co-circom-types::parse_field`.

**Review by dkales — APPROVED** [review]

**Review by 0xThemis — APPROVED** [review]
